D/D.sublime-syntax

```diff
@@ -8,339 +8,2057 @@ file_extensions:
+    - match: '\.(?!\.\.)'
       scope: storage.type.d
-    - match: \b(try|catch|finally|throw)\b
-      scope: keyword.control.exception.d
-    - match: \b(return|break|case|continue|default|do|while|for|switch|if|else)\b
-      scope: keyword.control.d
-    - match: \b(if|else|switch|iftype)\b
+      set:
+        - match: '\b({{name}})\b'
+          scope: storage.type.d
+          set: basic-type2
+    - match: '(?=\S)'
+      pop: true
+  basic-type2-after-brackets:
+    - match: '\]'
+      scope: punctuation.section.brackets.end.d
+      pop: true
+    - include: not-whitespace-illegal-pop
+  basic-type2-after-parens:
+    - match: '\)'
+      scope: punctuation.section.parens.end.d
+      pop: true
+    - include: not-whitespace-illegal-pop
+
+  typeof-value:
+    - match: '\breturn\b'
+      scope: keyword.other.d
+      pop: true
+    - include: value
+
+  function-type:
+    - match: (?=\()
+      set: [function-type-after, function-type-argument-definition-list]
+    - include: not-whitespace-illegal-pop
+  function-type-after:
+    - include: function-attribute-in
+    - match: '(?=\S)'
+      pop: true
+
+  # Function type arguments are slightly different to regular function
+  # arguments. With regular arguments a standalone "name" is asssumed to be a
+  # template argument, but with function types they must be a type.
+  function-type-argument-definition-list:
+    - match: '\('
+      scope: punctuation.section.group.begin.d
+      set:
+        - match: '\)'
+          scope: punctuation.section.group.end.d
+          pop: true
+        - match: '(?=\S)'
+          set: [function-type-argument, function-type-argument-type]
+    - include: not-whitespace-illegal
+  function-type-argument-type:
+    - include: parameter-attribute-in
+    - match: '\.\.\.'
+      scope: keyword.operator.variadic.d
+      pop: true
+    - include: type
+  function-type-argument:
+    - match: '\b{{name}}\b'
+      scope: variable.parameter.d
+      set:
+        - match: '\.\.\.'
+          scope: keyword.operator.variadic.d
+        - match: ','
+          scope: punctuation.separator.sequence.d
+          set: [function-type-argument, function-type-argument-type]
+        - match: '='
+          scope: keyword.operator.assignment.d
+          push: value
+        - match: ':'
+          scope: keyword.operator.assignment.d
+          push: value
+        - match: '\)'
+          scope: punctuation.section.group.end.d
+          pop: true
+        - include: not-whitespace-illegal
+    - match: ','
+      scope: punctuation.separator.sequence.d
+      push: function-type-argument-type
+    - match: '\)'
+      scope: punctuation.section.group.end.d
+      pop: true
+    - include: not-whitespace-illegal
+
+  function-or-variable-after-type:
+    - match: '\b({{operator_overloads}})\s*(?=\(|$)'
+      captures:
+        1: keyword.operator.overload.d
+      set: function-definition-begin
+    - match: '\b({{name}})\s*(?=\(|$)'
+      captures:
```

Comment:
The function definition (and any declarations) should have `meta.function` applied at least for the keyword, function name, parameters and return type. If you want to, you can continue the scope until the closing brace for the function body. For parameter, the scope should switch from `meta.function` to `meta.function.parameters`. For the return type, it should switch from `meta.function` to `meta.function.return-type`. Note that the scope should be swapped out, and not stacked.

Reply:
I'll get the others done, but `return-type` is impossible to do properly as the parser doesn't know whether the first thing on a line is a value or a type for either a variable or function until later, eg:

```
foo!(1, "bar")
// This would need meta.function.return-type, but we can't backtrace to this line to do that
bar() {}
```

Reply:
In situations like this, we typically use a heuristic and/or best effort.

If we see `int main()` we can set the appropriate meta scopes. If someone writes D that is hard to highlight when highlighting a single line at a time, then they'll just get a slightly less correct parsing.